enhancements/local-storage/automatic-detection.md

```diff
@@ -64,40 +65,299 @@ and we can use that information to sort devices into storage classes based on th
+	// Loop type device
+	Loop DeviceType = "loop"
+)
+
+type DiscoveredDevice struct {
+	// Name represents the persitent name of the device. For eg, /dev/disk/by-id/...
+	Name string `json:"name"`
+	// Model of the detected device
+	Model string `json:"model,omitempty"`
+	// Type of the detected device
+  Type DeviceType `json:"type"`
+	// Vendor of the detected device
+	Vendor string `json:"vendor,omitempty"`
+	// Size of the detected device
+	Size resource.Quantity `json:"size,omitempty"`
+	// Property represents whether the device type is rotational or not
+	Property DeviceMechanicalProperty `json:"property"`
+        // FSType represents the filesystem available on the device
+	FSType string `json:"fstype",omitempty`
+}
+
+type LocalVolumeDiscoveryResultSpec struct {
+  // NodeName represent the node for which LocalVolumeDiscoveryResult was created
+	NodeName string `json:"nodeName"`
+}
+
+type LocalVolumeDiscoveryResultStatus struct {
+	// Conditions is a list of conditions and their status.
+	Conditions          []operatorv1.OperatorCondition `json:"conditions,omitempty"`
+	DiscoveredTimeStamp string                         `json:"discoveredTimeStamp,omitempty"`
+	// Devices contains the list of devices on which LSO
+	// is capable of creating LocalPVs
+	// The devices in this list qualify these following conditions.
+	// - it should be a non-removable device.
+	// - it should not be a read-only device.
+	// - it should not be mounted anywhere
+	// - it should not be a boot device
+	// - it should not have child partition in lsblk output
+	DiscoveredDevices []DiscoveredDevice `json:"discoveredDevices"`
+}
+
+type LocalVolumeDiscoveryResult struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoveryResultSpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryResultStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoveryResultList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeDiscoveryResult `json:"items"`
+}
+
+```
+
+### Workflow of `Auto Device Discovery`
+The discovery of local devices will be done by the `LocalVolumeDiscovery` and `LocalVolumeDiscoveryResult` CR.
+
+#### LocalVolumeDiscovery Controller
+- Once the `LocalVolumeDiscovery` CR is deployed, its controller will:
+  - Create the `LocalVolumeDiscoveryResult` CRD if its not available.
+  - Create a daemonSet to run on each of the valid nodes.(Information about the valid list of nodes would be available in the `LocalVolumeDiscovery` CR as `nodeSelector`)
+  - The `Daemonset` should have `LocalVolumeDiscovery` CR as owner reference.
+  - Controller would remember the current timestamp when the discovery was triggered. (say `discoveryStarted`)
+  - Update the `Phase` in the `LocalVolumeDiscovery` CR to `Discovering`.
+- The DaemonSet pod running on each valid node will:
+  - Create `LocalVolumeDiscoveryResult` CR on that node if not present.
```

Comment:
No need to create the CR until after the list of devices are discovered the first time. During subsequent discoveries, the CR will just be updated whenever the device list changes.